### PR TITLE
Applab: New debug console layout

### PR DESCRIPTION
**Pull request by islemaster**
Introduces a new layout for the applab debug area that has a header and somewhat mirrors the toolbox/workspace motif being used in the code mode/design mode workspaces.  This seemed like a good stable changeset to stop and merge.

**Before:**
screenshot from 2015-06-18 12 52 22

**After:**
screenshot from 2015-06-18 13 05 25

I'm working from the following mock, but have done some modification of my own as well:
060115_codemode_flow2
# Key Changes
- Adds a resize bar with ellipsis above the debug area (formerly the row of buttons was the resize bar)
- Gives the debug area a header with labels
- Moves the debugging controls into a column to the left of the debug area.
- Adds a "clear" button for the debug console scrollback.

To be implemented in a future PR:
- Expand/collapse control for debug area.
- Maybe extract all the logic that manages state of the debug controls/area out of applab.js and into a helper.
## Running/Paused indicator

stopped_running_paused
- I moved the "running" spinner from the right end of the speed slider into the debug controls header - its purpose correlates better with the actions you take using those controls than with the actual execution speed.
- I replaced the existing spinner (which didn't scale well) with a CSS3-animated font-awesome spinner.
- I added a "paused" icon to differentiate the "not running" state from the "paused/broken" state.  I think it communicates better now.

### CODE REVIEW

**Comment by davidsbailey** [comment]
High-level comments (@sfilman may have an opinion too): 
1. I like the fully justified look you've chosen, rather than sizing individual buttons to their text.
2. orange is our primary call to action color. Maybe blue shown in the mocks is better for debugging?

**Comment by davidsbailey** [comment]
although actually, once they have broken into the debugger, the debugging buttons probably *are* the primary call to action.

File: apps/src/applab/applab.js [hunk 1]
```diff
@@ -588,7 +588,7 @@ Applab.init = function(config) {
     // Initialize the slider.
     var slider = document.getElementById('applab-slider');
     if (slider) {
-      Applab.speedSlider = new Slider(10, 27, 130, slider);
+      Applab.speedSlider = new Slider(10, 22, 130, slider);
```
Comment (breville) [inline apps/src/applab/applab.js]: can we document these magic numbers?

File: apps/src/applab/applab.js [hunk 2]
```diff
@@ -699,8 +703,8 @@ Applab.onMouseMoveDebugResizeBar = function (event) {
   //  we explicitly set bottom/height styles on the elements above, we need to
   //  do the same for these styles as well)
 
-  codeTextbox.style.minHeight = 'calc(50% - 21px)';
-  debugArea.style.maxHeight = 'calc(50% - 21px)';
+  codeTextbox.style.minHeight = 'calc(50% - 16px)';
```
Comment (breville) [inline apps/src/applab/applab.js]: is it weird to have these dimensions hardcoded as opposed to something in a central place?

**Comment by breville** [comment]
Code looks good.. just a couple little questions about constants.

Nice use of sass variables once again.

The presentation improvements are nice.  That font-awesome spinner is really neat, and it'd be good to use it across the board.

I find it slightly odd that step over has a curved arrow while step in & out use square/right-angled arrows.

I find the horizontal set of control buttons a little easier to parse than the three rows of buttons.  Partly this is because the groupings are difficult to get right.. i would have expected step in & step over to be paired up more than they are. It's common to alternate between the two, and having them opposite to each other in the grid feels a little awkward.

Also, is View Data in the same family of functions as the other four buttons?

**Comment by islemaster** [comment]
@davidsbailey I gave the "primary when broken" thing a shot and the first pass at least is hard on the eyes - since it briefly switches to a running state every time you step, the whole thing flashes blue a lot.

out

**Comment by islemaster** [comment]
@breville re: Parsing the button rows, maybe this will help:

The "View Data" button is there temporarily, it will move elsewhere soon because it's not really a debugging control.

Maybe we should reorder the buttons like this?
```
Continue   Step Over
Step Out   Step In
```

**Comment by breville** [comment]
I'm only a datapoint of one, but that ordering seems like it could help.

**Comment by islemaster** [comment]
Button colors settled: Blue, except for the "Continue" button:

out

**Comment by islemaster** [comment]
Here are the reordered buttons:

screenshot from 2015-06-18 14 21 21

File: apps/src/applab/applab.js (hunk 2 above)
Comment (islemaster) [inline apps/src/applab/applab.js]: It is weird, but the comment above reads:

```javascript
  // (we would do this in CSS, but for precedence rules to work properly, if
  //  we explicitly set bottom/height styles on the elements above, we need to
  //  do the same for these styles as well)
```

Perhaps these properties _should_ be in CSS, marked `!important`?

Comment (breville) [inline apps/src/applab/applab.js]: I don't feel super strongly.  That said, is 21px half of the height?  I can intuit it, but somebody arriving at the code in a couple years' time might be a bit confused.

Comment (islemaster) [inline apps/src/applab/applab.js]: Does [this change](https://github.com/code-dot-org/code-dot-org/commit/a9df1548eabcb2e49a2f347515577ad23859ef1e) clear things up?

**Pull request merged**